extensions/azure/cosmos/policy-store-cosmos/src/main/java/org/eclipse/dataspaceconnector/contract/definition/store/CosmosPolicyStore.java

```diff
@@ -0,0 +1,143 @@
+/*
+ *  Copyright (c) 2020 - 2022 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.contract.definition.store;
+
+import net.jodah.failsafe.RetryPolicy;
+import net.jodah.failsafe.function.CheckedSupplier;
+import org.eclipse.dataspaceconnector.azure.cosmos.CosmosDbApi;
+import org.eclipse.dataspaceconnector.common.concurrency.LockManager;
+import org.eclipse.dataspaceconnector.contract.definition.store.model.PolicyDocument;
+import org.eclipse.dataspaceconnector.policy.model.Policy;
+import org.eclipse.dataspaceconnector.spi.policy.store.PolicyStore;
+import org.eclipse.dataspaceconnector.spi.query.QueryResolver;
+import org.eclipse.dataspaceconnector.spi.query.QuerySpec;
+import org.eclipse.dataspaceconnector.spi.query.ReflectionBasedQueryResolver;
+import org.eclipse.dataspaceconnector.spi.types.TypeManager;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static net.jodah.failsafe.Failsafe.with;
+
+/**
+ * Implementation of the {@link PolicyStore} based on CosmosDB. This store implements simple write-through
+ * caching mechanics: read operations (e.g. findAll) always hit the cache, while write operations affect both the cache AND the
+ * database.
+ */
+public class CosmosPolicyStore implements PolicyStore {
+    private final CosmosDbApi cosmosDbApi;
+    private final TypeManager typeManager;
+    private final RetryPolicy<Object> retryPolicy;
+    private final LockManager lockManager;
+    private final String partitionKey;
+    private final QueryResolver<Policy> queryResolver;
+    private AtomicReference<Map<String, Policy>> objectCache;
```

Comment:
In case the idea is to be able to run two connectors in parallel, a local cache would not be updated when a request is send to the other connector and could lead to unexpected side effect.

Reply:
yup. that's what the `reload()` method is there for. in a clustered environment there probably would/should be a watchdog job to periodically refresh the cache, or a more sophisticated, possibly event-based mechanism.

Reply:
I don't know where you got that from. Many things are not yet implemented, such as a generic, consistent and possibly distributed caching strategy (across all data backends).